fix(ivy): do not always accept `undefined` for directive inputs

JoostK opened the pull request:
**fix(ivy): do not always accept `undefined` for directive inputs**

Prior to this change, the template type checker would always allow a
value of type `undefined` to be passed into a directive's inputs, even
if the input's type did not allow for it. This was due to how the type
constructor for a directive was generated, where a `Partial` mapped
type was used to allow for inputs to be unset. This essentially
introduces the `undefined` type as acceptable type for all inputs.

This commit removes the `Partial` type from the type constructor, which
means that we can no longer omit any properties that were unset.
Instead, any properties that are not set will still be included in the
type constructor call, having their value assigned to `any`.

Before:

```typescript
class NgForOf<T> {
  static ngTypeCtor<T>(init: Partial<Pick<NgForOf<T>,
    'ngForOf'|'ngForTrackBy'|'ngForTemplate'>>): NgForOf<T>;
}

NgForOf.ngTypeCtor(init: {ngForOf: ['foo', 'bar']});
```

After:

```typescript
class NgForOf<T> {
  static ngTypeCtor<T>(init: Pick<NgForOf<T>,
    'ngForOf'|'ngForTrackBy'|'ngForTemplate'>): NgForOf<T>;
}

NgForOf.ngTypeCtor(init: {
  ngForOf: ['foo', 'bar'],
  ngForTrackBy: null as any,
  ngForTemplate: null as any,
});
```

This change only affects generated type check code, the generated
runtime code is not affected.

Fixes #32690
Resolves FW-1606

---

**feat(ivy): check regular attributes that correspond with directive inputs**

Prior to this change, a static attribute that corresponds with a
directive's input would not be type-checked against the type of the
input. This is unfortunate, as a static value always has type `string`,
whereas the directive's input type might be something different. This
typically occurs when a developer forgets to enclose the attribute name
in brackets to make it a property binding.

This commit lets static attributes be considered as bindings with string
values, so that they will be properly type-checked.

---

**feat(ivy): disable strict null checks for input bindings**

This commit introduces an internal config option of the template type
checker that allows to disable strict null checks of input bindings to
directives. This may be particularly useful when a directive is from a
library that is not compiled with `strictNullChecks` enabled.

Right now, strict null checks are enabled when  `fullTemplateTypeCheck`
is turned on, and disabled when it's off. In the near future, several of
the internal configuration options will be added as public Angular
compiler options so that users can have fine-grained control over which
areas of the template type checker to enable, allowing for a more
incremental migration strategy.

JoostK committed:
fix(ivy): do not always accept `undefined` for directive inputs

Prior to this change, the template type checker would always allow a
value of type `undefined` to be passed into a directive's inputs, even
if the input's type did not allow for it. This was due to how the type
constructor for a directive was generated, where a `Partial` mapped
type was used to allow for inputs to be unset. This essentially
introduces the `undefined` type as acceptable type for all inputs.

This commit removes the `Partial` type from the type constructor, which
means that we can no longer omit any properties that were unset.
Instead, any properties that are not set will still be included in the
type constructor call, having their value assigned to `any`.

Before:

```typescript
class NgForOf<T> {
  static ngTypeCtor<T>(init: Partial<Pick<NgForOf<T>,
    'ngForOf'|'ngForTrackBy'|'ngForTemplate'>>): NgForOf<T>;
}

NgForOf.ngTypeCtor(init: {ngForOf: ['foo', 'bar']});
```

After:

```typescript
class NgForOf<T> {
  static ngTypeCtor<T>(init: Pick<NgForOf<T>,
    'ngForOf'|'ngForTrackBy'|'ngForTemplate'>): NgForOf<T>;
}

NgForOf.ngTypeCtor(init: {
  ngForOf: ['foo', 'bar'],
  ngForTrackBy: null as any,
  ngForTemplate: null as any,
});
```

This change only affects generated type check code, the generated
runtime code is not affected.

Fixes #32690
Resolves FW-1606

JoostK committed:
feat(ivy): check regular attributes that correspond with directive inputs

Prior to this change, a static attribute that corresponds with a
directive's input would not be type-checked against the type of the
input. This is unfortunate, as a static value always has type `string`,
whereas the directive's input type might be something different. This
typically occurs when a developer forgets to enclose the attribute name
in brackets to make it a property binding.

This commit lets static attributes be considered as bindings with string
values, so that they will be properly type-checked.

JoostK committed:
fixup! fix(ivy): do not always accept `undefined` for directive inputs

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -817,28 +825,52 @@ function tcbGetInputBindingExpressions(
-        expression: expr,
-        sourceSpan: attr.sourceSpan,
-      });
+      expr = tcbExpression(attr.value, tcb, scope, attr.valueSpan || attr.sourceSpan);
```
I had a discussion with @kara about some concerns @jelbourn has here. Specifically, libraries today may be built without `strictNullChecks` enabled and with typings that don't explicitly accept `undefined` (or `null` for that matter), which could cause some pain for users as a result of the extra strictness here.

Let's introduce a `TypeCheckingConfig` flag `strictNullInputBindings`, set to `true` by default. If `false`, let's wrap `expr` here in a `ts.NonNullExpression` for compatibility with broken library typings. Then we'll be all set to expose this flag if we decide we need to give users control here.

alxhub reviewed (approved):
LGTM, with one comment.

JoostK committed:
feat(ivy): disable strict null checks for input bindings

This commit introduces an internal config option of the template type
checker that allows to disable strict null checks of input bindings to
directives. This may be particularly useful when a directive is from a
library that is not compiled with `strictNullChecks` enabled.

Right now, strict null checks are enabled when  `fullTemplateTypeCheck`
is turned on, and disabled when it's off. In the near future, several of
the internal configuration options will be added as public Angular
compiler options so that users can have fine-grained control over which
areas of the template type checker to enable, allowing for a more
incremental migration strategy.

JoostK committed:
build(docs-infra): fix template type check errors

In #33066 a limitation of Ivy's template type checker was fixed, where
all directive inputs would incorrectly allow `undefined` to be passed,
even when the input's type did not allow for it. Due to the fix, some
additional type errors were uncovered in AIO, where potential
`undefined` values would be passed to inputs that were not typed to
allow `undefined`.

JoostK committed:
fixup! feat(ivy): check regular attributes that correspond with directive inputs

gkalpak commented on aio/src/app/custom-elements/code/code-tabs.component.ts:
```diff
@@ -78,7 +78,7 @@ export class CodeTabsComponent implements OnInit, AfterViewInit {
       header: tabContent.getAttribute('header') || undefined,
       language: tabContent.getAttribute('language') || undefined,
-      linenums: tabContent.getAttribute('linenums') || this.linenums,
+      linenums: tabContent.getAttribute('linenums') || this.linenums || false,
```
Currently, `undefined` ends up behaving the same as `false`, but this wasn't always the case. `undefined` means "use the default value", which might change in the future. Can we pass `undefined` through to the `CodeComponent`, so that we have one place where `undefined` it converted to a default value?

gkalpak reviewed (approved):
LGTM for `docs-infra` 👍

kara commented:
[VE presubmit](https://test.corp.google.com/ui#id=OCL:274613159:BASE:274613203:1571075670354:9e733e3c)

[Ivy presubmit](http://test/OCL:274613159:BASE:274613567:1571075765291:7124c272)

Splaktar commented:
> This commit lets static attributes be considered as bindings with string
values, so that they will be properly type-checked.

After updating to `next.11` with this change, I'm getting hundreds of these template type checking errors:
```ts
error TS2322: Type 'string' is not assignable to type 'boolean'.
```
This can be for simple things like `<button mat-raised-button disabled>disabled</button>`. The specific error in this case is for the `disabled` attribute.

It also occurs for things like this
```html
    <div
      cdkDropList
      [cdkDropListData]="items"
      class="example-list"
      cdkDropListSortingDisabled
      (cdkDropListDropped)="drop($event)">
```
Where `cdkDropListSortingDisabled` is giving the error and it maps to a setter that has a `boolean` type.

Is this related to this change? Or should I keep looking through other changes in `next.11` that would have caused this?

JoostK commented:
@Splaktar oh no, that was definitely not the intention of that commit (although it is the one that introduces your type check errors, indeed)

The idea behind checking static attributes is for situations like these:

```html
<my-cmp mode="drak" config="config"></my-cmp>
```

with `my-cmp` binding to the following component class:

```typescript
export class MyCmp {
  @Input() mode: 'light' | 'dark';
  @Input() config: SomeConfig;
}
```

In this case, the template has an invalid binding for `mode` (it's misspelled) and the `config` attribute should have been a binding instead of a static attribute:

```html
<my-cmp mode="dark" [config]="config"></my-cmp>
```

This is why static attributes now get checked as well.

In a pedantic world, the errors you're seeing for `disabled` are technically correct, but highly undesirable. There's work planned to introduce public facing strictness flags to be able to influence the behavior of the type checker here, unfortunately that is not in `next.11`. This is my mistake, the change related to attributes should have landed together with having these options available as it would allow to easily disable them. My apologies!

The scale of this change was not caught as I realize now that the Material unit tests we run have full template type check disabled.

Splaktar commented:
@JoostK OK, that's good to know, thank you for that information!

FW-1475 has a pretty good explanation about the kinds of ergonomics that Angular Material and other libraries depend upon, especially around `@Input`s with `boolean` and `number` types and things like `required` and `disabled`.
